### ISSUE TITLE
[QUESTION] Bootstrap-style submenu's

### ISSUE THREAD

**Issue by thijskok** [spatie/laravel-menu#4]
- [x] I have read and understood the [contributors guide]().
- [X] I have checked the pull requests tab for existing solutions/implementations to my issue/suggestion.
- [X] I have checked that the bug-fix I am reporting can be replicated.

---

### Description of the problem

I am struggling to get submenu's working with Bootstrap 3 navbar's. It simply boils down to this: the examples you provide generate submenu's in a separate `<li>`, while Bootstrap expects you to add a submenu right after the anchor, within the same `<li>` of the parent.

Laravel-menu generates:
```html
<ul>
    <li>
        <a href="/">Menu</a>
    </li>
    <li>
        <ul>
            <li>
                <a href="/basic-usage/your-first-menu">Your First Menu</a>
            </li>
        </ul>
    </li>
</ul>
```

While Bootstrap expects:
```html
<ul>
    <li>
        <a href="/">Menu</a>
        <ul>
            <li>
                <a href="/basic-usage/your-first-menu">Your First Menu</a>
            </li>
        </ul>
    </li>
</ul>
```

In order to generate the second code fragment using laravel-menu, I would expect something like this in PHP:

```php
Menu::new()->add(Link::to('/', 'item'), Menu::new()
    ->add(Link::to('/sub', 'subitem'))
);
```

Is there something I missed? Or are there any plans in making this work, making Bootstrap submenu's possible?

### Steps to reproduce

I have used the [submenu code examples](https://docs.spatie.be/menu/v1/basic-usage/adding-items) from the documentation.

**Comment by sebastiandedeyne** [comment]
Hmm this isn't ideal. I'm going to give it some thought and push an elegant solution in the next few days.

In the mean time, you could go about it like this, but the parent item won't ever be set active this way:

```php
Menu::new()
    ->add(Menu::new()
        ->prepend(Link::to('/', 'item')->render())
        ->add(Link::to('/sub', 'subitem'))
    );

**Comment by agentphoenix** [comment]
Adding on to this issue, when using `prepend` I can't actually attach a class to the parent item, so even with this, I haven't been able to get the rendered output the way I need them. Here's what I'm using to build the menu:

```php
// Build the header for the sub menu
$header = LinkBuilder::to('#', $mainMenuItem->title.' <span class="caret"></span>');

// Start building the submenu and prepend the first item
$submenu = MenuBuilder::new()->addClass('dropdown-menu');
$submenu->prepend($header->addClass('dropdown-toggle')->addParentClass('dropdown'));
```

Given the above code, I'd expect to see the `li` with a class of _dropdown_, but it doesn't actually have any class:

```html
<li><a href="#" class="dropdown-toggle">Manage <span class="caret"></span></a>
    <ul class="dropdown-menu">
        ...
    </ul>
</li>
```

What I would expect to see given the PHP above is:

```html
<li class="dropdown"><a href="#" class="dropdown-toggle">Manage <span class="caret"></span></a>
    <ul class="dropdown-menu">
        ...
    </ul>
</li>
```

Am I missing something or is attaching to the parent not possible when prepending?

**Comment by danhunsaker** [comment]
Have you tried adding the parent class to the submenu itself, rather than the prepended header?

```php
// Build the header for the sub menu
$header = LinkBuilder::to('#', $mainMenuItem->title.' <span class="caret"></span>');

// Start building the submenu and prepend the first item
$submenu = MenuBuilder::new()->addClass('dropdown-menu')->addParentClass('dropdown');
$submenu->prepend($header->addClass('dropdown-toggle'));
```

**Comment by agentphoenix** [comment]
Worked like a charm. That's probably the only thing I didn't try when I was messing around with all of this. Appreciate this, @danhunsaker!

**Comment by danhunsaker** [comment]
Happy to help!

**Comment by sebastiandedeyne** [comment]
I'm planning to add a `submenu` method to the next release (probably tomorrow already), here's what the syntax would look like:

```php
// Header and menu argument
Menu::new()->submenu('Hi', Menu::new()->link('#', 'Yo'));
```

```html
<ul>
    <li>
        Hi
        <ul>
            <li><a href=#>Yo</a></li>
        </ul>
    </li>
</ul>
```


```php
// Header can also be an `Item`
Menu::new()->submenu(Link::to('#', 'Hi'), Menu::new()->link('#', 'Yo'));
```

```html
<ul>
    <li>
        <a href=#>Hi</a>
        <ul>
            <li><a href=#>Yo</a></li>
        </ul>
    </li>
</ul>
```

```php
// Header isn't required
Menu::new()->submenu(Menu::new()->link('#', 'Yo'));
```

```html
<ul>
    <li>
        <ul>
            <li><a href=#>Yo</a></li>
        </ul>
    </li>
</ul>
```

```php
// If the menu argument is a callable, a new menu instance will be passed through with the same filters as the previous one, which allows cascading link prefixes.
Menu::new()->prefixLinks('/foo')->submenu(function (Menu $menu): Menu {
    $menu->link('/bar', 'Bar');
});
```

```html
<ul>
    <li>
        <ul>
            <li><a href="/foo/bar">Bar</a></li>
        </ul>
    </li>
</ul>
```

**Comment by danhunsaker** [comment]
That actually looks pretty great! Looking forward to it.

**Comment by sebastiandedeyne** [comment]
Released `spatie/menu` v1.2 yesterday 🎉 Should come through if you run composer update!

https://github.com/spatie/menu/releases/tag/1.2.0

I also added a Bootstrap menu example in the docs, using some of the menu's new features: https://docs.spatie.be/menu/v1/examples/bootstrap-menu
